### ISSUE TITLE
Incorrect reportUndefinedVariable for builtin __qualname__

### ISSUE THREAD

**Issue by RMBeristain** [microsoft/pylance-release#1376]
Pylance mistakenly raises a reportUndefinedVariable warning for dundermethod__qualname__

## Environment data

-   Language Server version: v2021.5.4
-   OS and version: 
    - Windows 10 Version	10.0.18363 Build 18363
    - WSL Ubuntu 20.04.2 LTS (Linux VAU-NB-11514 4.4.0-18362-Microsoft #1049-Microsoft Thu Aug 14 12:01:00 PST 2020 x86_64 x86_64 x86_64 GNU/Linux )
-   Python version (& distribution if applicable, e.g. Anaconda): cPython 3.8.5

## Expected behaviour

Pylance should recognise __qualname__ as valid dundermethod in a class defintion and not raise any warning.

Example:

```python
class Test:
    "Pylance bug"
    my_name = __qualname__

>>> t = Test()
>>> t.my_name
'Test'
```

## Actual behaviour

Pylance raises reportUndefinedVariable  for the above code.

```
"__qualname__" is not defined
```

## Logs
```
{
	"resource": "/mnt/c/Users/virtualraider/Documents/Code/GitLab/deleteme.py",
	"owner": "_generated_diagnostic_collection_name_#1",
	"code": {
		"value": "reportUndefinedVariable",
		"target": {
			"$mid": 1,
			"external": "https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules",
			"path": "/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "diagnostic-severity-rules"
		}
	},
	"severity": 4,
	"message": "\"__qualname__\" is not defined",
	"source": "Pylance",
	"startLineNumber": 6,
	"startColumn": 15,
	"endLineNumber": 6,
	"endColumn": 27
}
```

**Comment by erictraut** [comment]
The documentation for `__qualname__` within a class does not suggest that it is accessible from within the class body itself. See [PEP 3155](https://www.python.org/dev/peps/pep-3155/) for the introduction of this attribute, and note that none of its examples show it being used within the class body. That said, I've confirmed that the Python interpreter does allow this. I wonder whether this was intentional or an oversight. It appears to be somewhat inconsistent because some dundered attributes like `__doc__` and `__module__` are accessible within the class body but others like `__dict__` and `__mro__` are not. Perhaps that's because the latter are still being constructed at the time the class body is executed.

The same PEP introduced `__qualname__` for functions, and it is not accessible from within a function body (it generates a NameError at runtime).

Mypy (another popular type checker) also reports the use of `__qualname__` and `__module__` as errors.

So I wonder whether this is intended behavior or just an implementation detail that you should not be relying on.

Do you happen to know of any official Python documentation that explains that the use of `__qualname__` within a class body is allowed?

**Comment by RMBeristain** [comment]
Hi @erictraut , thanks for replying so quick!

I do not have any official documentation other than that PEP, sorry.

I have seen it in the dataclass implementation, which is where I found out about it 

https://github.com/python/cpython/blob/a6a20658814e8668966fc86de0e80a4772864781/Lib/dataclasses.py#L582

I wanted to use it to build a qualified name hierarchy for logging so instead of:

```
class MyClass:
  log = module_logger.getChild("MyClass")
```

I can do 
```
class MyClass:
  log = module_logger.getChild(__qualname__)
```

and this future-proofs any code refactoring of the module name.

**Comment by erictraut** [comment]
I've updated the logic in pyright/pylance to handle this case consistently with the Python interpreter. I also filed a bug report for mypy, since it contains the same problem. After some experimentation, I found that `__module__` needs to be handled in the same manner as `__qualname__`.

This will be fixed in the next release of pylance. Thanks for reporting the problem.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.6.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202160-2-june-2021

**Comment by maresb** [comment]
@erictraut, there is unfortunately a regression of this bug since `v2021.8.1`. Could we please reopen this issue?

**Comment by erictraut** [comment]
`__qualname__` is very strange. It doesn't show up in the `__dict__` for a class. If you call `dir()` for the class, it isn't listed. When you call `locals()` within a class body, it is listed. So it appears to be a local variable injected into the scope at the time the class body is executed.

I'll note that mypy doesn't handle this correctly either. It's not surprising given that it appears to be a very strange special case.

I've added some logic that makes this work. This will be included in the next release.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.12.1, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2021121-9-december-2021)
